compiler/rustc_ty_utils/src/instance.rs

```diff
@@ -55,6 +55,46 @@ fn resolve_instance<'tcx>(
                 debug!(" => trivial drop glue");
                 ty::InstanceDef::DropGlue(def_id, None)
             }
+        } else if Some(def_id) == tcx.lang_items().async_drop_in_place_fn() {
+            let ty = args.type_at(0);
+
+            match *ty.kind() {
+                ty::Array(..)
+                | ty::Slice(_)
+                | ty::Tuple(_)
+                | ty::Bool
+                | ty::Char
+                | ty::Int(_)
+                | ty::Uint(_)
+                | ty::Float(_)
+                | ty::Str
```

Comment:
why are all these accepted here in contrast to `drop_in_place_fn`?

Reply:
This code fixed a bug. I presume it was because `needs_drop` condition filters out scalar types, which we don't have for async drop. And no, `needs_async_drop` does not make sense to me because of structure of async drop API from core (at least for now).

Reply:
hmm... I find it odd that we have to invoke async drop glue on types that are `Copy`. I guess if it's a simpler implementation for now that may be ok, but it should definitely have a bunch of FIXMEs and a note in the tracking issue to remove it later.

Reply:
This bug was essentially about the compiler trying to find `AsyncDrop` impl for scalar types like integers.

Reply:
hold up. this is `AsyncDrop`, there needs to be no `AsyncDrop` impl for these. the `Empty` shim of `AsyncDestruct` will not invoke it after all.

Reply:
what caused the solver to look for these impls?

Reply:
> what caused the solver to look for these impls?

I have copied drop glue branch above and removed if with it's else branch due to not having `needs_async_drop` right now implemented (I haven't decided how it should function yet). So it was basically filtering out scalar types with `_ => return Ok(None),` remains of which you can see in the last match arm.

Reply:
> why are all these accepted here in contrast to `drop_in_place_fn`?

Actually I've changed this code to be similar to regular drop's analogous code. Basically any type that have `Noop` async destructor to instantiate `InstanceDef::AsyncDropGlueCtorShim(def_id, None)` in 24a24ec6ba24bfe5e0980d22f585c98a608ec701...e239e73a77cda38b53e9a7716e01f0c2af995ff5.